### PR TITLE
Add extractor for VrtNU

### PR DESCRIPTION
**Pull request by mrBliss**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This PR address #11829 by adding a new extractor for VrtNU to `vrt.py`. The extractor is based on the existing extractor for canvas. As all videos require authentication the sole test is skipped. I have successfully used this extractor for a number of different videos (using `--cookies` for authentication).

I welcome all feedback.

### CODE REVIEW

**Comment by mrBliss** [comment]
I don't mind squashing the commits together if that is preferred.

**Comment by dstftw** [comment]
This is the copy paste of `CanvasIE`. You should generalize common code instead of duplicating it.

**Comment by mrBliss** [comment]
True, all the code about `formats` and `subtitles` is copied from `CanvasIE`. However, the extraction of the info from the webpage is new. I thought at first about extending `CanvasIE`, but that would lead to many special cases for VrtNU, which does not help the readability. 

So how should the common code be generalised? By subclassing or by extracting it into methods that are put in `common.py` or in `canvas.py` as static methods or functions?

To me the two candidates for extraction seem to be:

```python
formats = []
for target in data['targetUrls']:
    format_url, format_type = target.get('url'), target.get('type')
    if not format_url or not format_type:
        continue
    if format_type == 'HLS':
        formats.extend(self._extract_m3u8_formats(
            format_url, display_id, entry_protocol='m3u8_native',
            ext='mp4', preference=0, fatal=False, m3u8_id=format_type))
    elif format_type == 'HDS':
        formats.extend(self._extract_f4m_formats(
            format_url, display_id, f4m_id=format_type, fatal=False))
    elif format_type == 'MPEG_DASH':
        formats.extend(self._extract_mpd_formats(
            format_url, display_id, mpd_id=format_type, fatal=False))
    else:
        formats.append({
            'format_id': format_type,
            'url': format_url,
        })
self._sort_formats(formats)
```
and:
```python
subtitles = {}
subtitle_urls = data.get('subtitleUrls')
if isinstance(subtitle_urls, list):
    for subtitle in subtitle_urls:
        subtitle_url = subtitle.get('url')
        if subtitle_url and subtitle.get('type') == 'CLOSED':
            subtitles.setdefault('nl', []).append({'url': subtitle_url})
```

Another hard problem to solve is to come up with sensible names for these methods :smiley:

**Comment by dstftw** [comment]
You should not pollute `common.py` with particular extractor specific code. Easiest way is to extract common code in a base class.

**Comment by mrBliss** [comment]
Another try, but now as a subclass of `CanvasIE`.

File: youtube_dl/extractor/canvas.py [hunk 1]
```diff
@@ -117,3 +125,63 @@ def _real_extract(self, url):
...
+        securevideo_url = clean_url + '.securevideo.json'
+
+        json = self._download_json(securevideo_url, display_id)
+        # There is only one entry, but with an unknown key, so just get the
+        # first one
+        video_id = list(json.values())[0].get('mzid')
+
+        data = self._download_json(
+            'https://mediazone.vrt.be/api/v1/vrtvideo/assets/%s'
+            % video_id, display_id)
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: 170-172 - code duplication.

File: youtube_dl/extractor/canvas.py [hunk 2]
```diff
@@ -117,3 +125,63 @@ def _real_extract(self, url):
...
+        return {
+            'id': video_id,
+            'display_id': display_id,
+            'title': title,
+            'description': description,
+            'formats': formats,
+            'duration': float_or_none(data.get('duration'), 1000),
+            'thumbnail': data.get('posterImageUrl'),
+            'subtitles': subtitles,
+        }
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: 178-187 - code duplication.

Comment (mrBliss) [inline youtube_dl/extractor/canvas.py]: Should I go for a method returning a tuple of `duration` and `thumbnail`? The SLOC will only increase though.

File: youtube_dl/extractor/canvas.py (hunk 1 above)
Comment (mrBliss) [inline youtube_dl/extractor/canvas.py]: Done

File: youtube_dl/extractor/canvas.py (hunk 2 above)
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: You should have only one single method for extracting info.

Comment (mrBliss) [inline youtube_dl/extractor/canvas.py]: Okay, I have now taken this approach and rebased again.

Comment (mrBliss) [inline youtube_dl/extractor/canvas.py]: @dstftw Have you had time to look at my latest changes yet? Thanks

**Comment by rubenv** [comment]
I've been testing this PR and it works great. The Travis failure seems to be related to a timeout on the Travis side of things and should probably be ignored (ore restarted).

Nice work @mrBliss!

**Comment by MichaelDeBoey** [comment]
Don't know if this can be merged @dstftw, but for what I'm seeing in this repo the Travis CI error(s) aren't related to @mrBliss' code

**Comment by MichaelDeBoey** [comment]
I don't know how you implemented this @mrBliss, but I've found a really easy way of getting the correct download-link for VrtNU :-)

1. You just have to add `.securevideo.json` to the link (`https://www.vrt.be/vrtnu/a-z/thuis/22/thuis-s22a4160` will be `https://www.vrt.be/vrtnu/a-z/thuis/22/thuis-s22a4160.securevideo.json`)
2. You'll get an `mzid` from that `.json`-file.
3. Use that `mzid` at the end of `https://mediazone.vrt.be/api/v1/vrtvideo/assets/` so you get: `https://mediazone.vrt.be/api/v1/vrtvideo/assets/md-ast-0fa95ed5-4541-4e3f-b436-ee552bd613fb_1488278903849`
4. Now you see all sorts (`hls`, `.mp4`, ...) of download possibilities

**Comment by mrBliss** [comment]
@MichaelDeBoey that's exactly how I've implemented it :smiley:

**Comment by MichaelDeBoey** [comment]
> @MichaelDeBoey that's exactly how I've implemented it 😃
OK nice :-)
I just found out this method on userbase (the forum where I also found the method for #11912), so wanted to let you know :-)

**Comment by MichaelDeBoey** [comment]
Any progress on this @mrBliss @dstftw?

**Comment by Invictaz** [comment]
@mrBliss I'm running your version and it does not work. See my screenshot

-Running Python 3.6.0

`C:\Users\Jurgen\Downloads\youtube-dl-vrtnu\youtube-dl-vrtnu\youtube_dl>python __
main__.py https://www.vrt.be/vrtnu/a-z/postbus-x/3/postbus-x-s3a8
[VrtNU] postbus-x-s3a8: Downloading webpage
[VrtNU] postbus-x-s3a8: Downloading JSON metadata
ERROR: Unable to download JSON metadata: HTTP Error 401: Unauthorized (caused by
 <HTTPError 401: 'Unauthorized'>); please report this issue on https://yt-dl.org
/bug . Make sure you are using the latest version; see  https://yt-dl.org/update
  on how to update. Be sure to call youtube-dl with the --verbose flag and inclu
de its complete output.`

**Comment by mrBliss** [comment]
@Invictaz the extractor doesn't yet support authentication, e.g., via `--username` and `--password`, so you are supposed to get that error.

My current workaround is to log in using my browser, export my cookies using the [Export Cookies add-on](https://addons.mozilla.org/en-gb/firefox/addon/export-cookies/) (a similar extension for Chrome undoubtedly exists), and pass them to youtube-dl via the `--cookies` flag. This has worked flawlessly for me for the past couple of months.

When I have some more time, I will try to add authentication.

**Comment by Invictaz** [comment]
@mrBliss it's working now. Not the easy method I hoped for but undoubtly that will come. Please include in your script a method to extract the .vtt subtitles and convert them to .srt. This is in for example https://www.vrt.be/vrtnu/a-z/bobbejaan/2015/bobbejaan-s2015/

They are in .vtt format, very similar to SRT. In the .securevideo.json is also the link to the subtitles.

Do not make them hardcoded but as a separate file please.

Example of your script working:

C:\Users\Downloads\youtube-dl-vrtnu\youtube-dl-vrtnu\youtube_dl>python __
main__.py --cookies cookies.txt https://www.vrt.be/vrtnu/a-z/postbus-x/3/postbus
-x-s3a8
[VrtNU] postbus-x-s3a8: Downloading webpage
[VrtNU] postbus-x-s3a8: Downloading JSON metadata
[VrtNU] postbus-x-s3a8: Downloading JSON metadata
[VrtNU] postbus-x-s3a8: Downloading m3u8 information
[VrtNU] postbus-x-s3a8: Downloading f4m manifest
[VrtNU] postbus-x-s3a8: Downloading MPD manifest
[hlsnative] Downloading m3u8 manifest
[hlsnative] Total fragments: 124
[download] Destination: De pyromaan-mz-ast-6a65c311-e8a0-4be8-a4ad-7ff0ee9a25b6_
1467042023805.mp4
[download]  84.7% of ~240.25MiB at  1.02MiB/s ETA 00:51

**Comment by Invictaz** [comment]
@mrBliss Oh and there is this

WARNING: mz-ast-6a65c311-e8a0-4be8-a4ad-7ff0ee9a25b6_1467042023805: malformated
aac bitstream. Install ffmpeg or avconv to fix this automatically.

**Comment by mrBliss** [comment]
@Invictaz 
My extractor already downloads the .vtt files if you pass the right flag, but I don't think it's the responsibility of the extractor to convert them to .srt. You can just use `ffmpeg` for this: `ffmpeg -i foo.vtt foo.srt`.

That warning you get is normal and not specific to my extractor. Just read the full warning, it literally says how you can fix it.

**Comment by mrBliss** [comment]
I added support for authentication via the `-u` and `-p` flags. Only e-mail accounts are supported though.

@MichaelDeBoey @Invictaz can you guys try this out with your accounts? Thanks

**Comment by rubenv** [comment]
@mrBliss Works like a charm, good stuff! 👍

**Comment by MichaelDeBoey** [comment]
@dstftw I think this one can be merged?

**Comment by midas02** [comment]
@mrBliss Can you check the following? I tried https://github.com/rg3/youtube-dl/pull/11873/commits/57fc66764242654892928a69d444467c7586fc95 (simplify login logic), and couldn't download http://www.vrt.be/vrtnu/a-z/reizen-waes/3/reizen-waes-s3a8-venezuela/. When accessing the link over httpS://, it did work! Tried release https://github.com/rg3/youtube-dl/pull/11873/commits/8317eeb261d1a69bc3db2493b1ec620191699c92, and that one didn't have the problem. But maybe I just made a mistake somewhere, please check.

**Comment by mrBliss** [comment]
@midas02 The S in httpS is not the problem. The authentication seems to be a bit flaky, sometimes it just works when you try again. I don't think my last commit is the cause because I have encountered the issue before. For me it works 8 out of 10 times.

I'm not sure how to fix this, because I can't reliably reproduce it. Maybe it's some kind of a timing issue? Maybe I need to add a retry somewhere in case of a failed authentication.

**Comment by midas02** [comment]
@mrBliss Ok. I checked again, and it did work. False alarm.

**Comment by MichaelDeBoey** [comment]
@dstftw I think this one can be merged?

**Comment by MichaelDeBoey** [comment]
Any progress on this one?

CC: @mrBliss @dstftw

File: youtube_dl/extractor/canvas.py [hunk 3]
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
...
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        display_id = mobj.group('id')
+
+        webpage = self._download_webpage(url, display_id)
+
+        title = self._html_search_regex(
+            r'<h1 class="content__heading">(.+?)</h1>',
+            webpage, 'title', default=None, flags=(re.M | re.S)).strip()
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: Consistently use flags inside regex.

File: youtube_dl/extractor/canvas.py [hunk 4]
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
...
+            [r'''(?xms)<div\ class="tabs__tab\ tabs__tab--active">\s*
+                    <span>seizoen\ (.+?)</span>\s*
+                </div>''',
+             r'<option value="seizoen (.+?)" data-href="[^"]+?" selected>'],
+            webpage, 'season', default=None)
+
+        season_number = int_or_none(season)
+
+        # Sometimes the season is the year, so don't use it as the season number
+        if season_number and season_number > 1000:
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: This can be expressed in regex.

File: youtube_dl/extractor/canvas.py [hunk 5]
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
...
+            'url': 'https://www.vrt.be/vrtnu/a-z/trapped/1/trapped-s1a4/',
+            'info_dict': {
+                'id': 'md-ast-01e7d193-63ab-4d89-9ab1-96b23315ebec_1489574042173',
+                'ext': 'mp4',
+                'title': 'Trapped',
+                'description': 'md5:6f4f779d176252f887fdd938910d6094',
+                'duration': 3011.17,
+                'thumbnail': r're:^https?://.*\.jpg$'
+            },
+            'skip': 'This video is only available for registered users'
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: Plays fine without any authentication in browser.

File: youtube_dl/extractor/canvas.py [hunk 6]
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
...
+    _APIKEY = '3_0Z2HujMtiWq_pkAjgnS2Md2E11a1AwZjYiBETtwNE-EoEHDINgtnvcAOpNgmrVGy'
+    _CONTEXT_ID = 'R1070628488'
+
+    def _real_initialize(self):
+        self._login()
+
+    def _login(self):
+        username, password = self._get_login_info()
+        if username is None:
+            self.raise_login_required()
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: This is not true at the moment.

File: youtube_dl/extractor/canvas.py [hunk 7]
```diff
@@ -110,10 +104,153 @@ def _real_extract(self, url):
...
+            data=json.dumps({
+                'uid': auth_info['UID'],
+                'uidsig': auth_info['UIDSignature'],
+                'ts': auth_info['signatureTimestamp'],
+                'email': auth_info['profile']['email'],
+            }).encode('utf-8'))
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        display_id = mobj.group('id')
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: `self._match_id`.

File: youtube_dl/extractor/canvas.py (hunk 5 above)
Comment (mrBliss) [inline youtube_dl/extractor/canvas.py]: I cannot reproduce this, I have to authenticate in order to view [the video](https://www.vrt.be/vrtnu/a-z/trapped/1/trapped-s1a4/) in a browser (Firefox, Chromium, Midori).

File: youtube_dl/extractor/canvas.py (hunk 4 above)
Comment (mrBliss) [inline youtube_dl/extractor/canvas.py]: If I do this with a regex and the season is something like for example "2017", both `season` and `season_number` will be `None`. Whereas previously, `season` would still be `"2017"`. For me the new behaviour is fine, as I don't care about non-numbered seasons.

**Comment by mrBliss** [comment]
I have pushed a new commit addressing your comments except the comment(s) related to authentication, as I cannot reproduce it.

**Comment by midas02** [comment]
@mrBliss Could you add some code to catch links terminating in ".relevant"? The VRT uses them to redirect to the latest episode of a series. Catching them would allow one to automatically download the latest available episode.
E.g.: https://www.vrt.be/vrtnu/a-z/sportweekend.relevant/
The thing is that the current logic doesn't work for them. Their json has a keyword "url" which holds the page to redirect to. So it would need to be detected first, and redirected.
I could analyse it and propose some code if you wouldn't have time for it.

**Comment by MichaelDeBoey** [comment]
That would indeed be a nice addition @midas02! :-)

But I think we could already merge this one and add it later?
Otherwise we keep adding extra features to the extractor without merging it to the master branch... 😞

**Comment by mrBliss** [comment]
@midas02 Easy to add, see the last commit

**Comment by mrBliss** [comment]
@dstftw can you look at this, if you have some time to spare?

**Comment by mrBliss** [comment]
@dstftw do you have some time to look at this again? I would like your opinion on my last commit (f07542c)

**Comment by brechtm** [comment]
The extractor is no longer working for me:

```
$ /Users/brechtm/Documents/Code/youtube-dl/youtube-dl --verbose --username ****** --password ****** https://www.vrt.be/vrtnu/a-z/allah-in-europa/2017/allah-in-europa-s2017a2-bosnie-hongarije/
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'--verbose', u'--username', u'PRIVATE', u'--password', u'PRIVATE', u'https://www.vrt.be/vrtnu/a-z/allah-in-europa/2017/allah-in-europa-s2017a2-bosnie-hongarije/']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.06.12
[debug] Python version 2.7.13 - Darwin-16.7.0-x86_64-i386-64bit
[debug] exe versions: ffmpeg 3.2.4, ffprobe 3.2.4
[debug] Proxy map: {}
[VrtNU] Logging in
[VrtNU] Requesting a token
[VrtNU] allah-in-europa-s2017a2-bosnie-hongarije: Downloading webpage
[VrtNU] allah-in-europa-s2017a2-bosnie-hongarije: Downloading JSON metadata
ERROR: Unable to download JSON metadata: HTTP Error 404: Not Found (caused by HTTPError()); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; type  youtube-dl -U  to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
  File "/Users/brechtm/Documents/Code/youtube-dl/youtube-dl/youtube_dl/extractor/common.py", line 502, in _request_webpage
    return self._downloader.urlopen(url_or_request)
  File "/Users/brechtm/Documents/Code/youtube-dl/youtube-dl/youtube_dl/YoutubeDL.py", line 2129, in urlopen
    return self._opener.open(req, timeout=self._socket_timeout)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 435, in open
    response = meth(req, response)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 548, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 473, in error
    return self._call_chain(*args)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/urllib2.py", line 556, in http_error_default
    raise HTTPError(req.get_full_url(), code, msg, hdrs, fp)
```

**Comment by mrBliss** [comment]
I have updated this PR in light of 117589dfa2ac32566dc705ee7e712139105e6dd8. Because of the refactoring in the aforementioned commit, rebasing became too complex, so I squashed the commits first.

@dstftw Can you take a look at this? I know of several people (myself included) that have been using this PR for over 6 months. It would be nice if it finally got merged.

File: youtube_dl/extractor/canvas.py [hunk 8]
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
...
+        json = self._download_json(securevideo_url, display_id)
+        # We are dealing with a '../<show>.relevant' URL
+        redirect_url = json.get('url')
+        if redirect_url:
+            return self._real_extract('https:' + redirect_url)
+        else:
+            # There is only one entry, but with an unknown key, so just get
+            # the first one
+            video_id = list(json.values())[0].get('videoid')
+            print(video_id)
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: All debug code should be removed.

File: youtube_dl/extractor/canvas.py [hunk 9]
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
...
+
+        auth_info = self._download_json(
+            'https://accounts.eu1.gigya.com/accounts.login', None,
+            note='Logging in', errnote='Unable to log in',
+            data=urlencode_postdata(auth_data))
+
+        error_message = auth_info.get('errorDetails') or auth_info.get('errorMessage')
+        if error_message:
+            raise ExtractorError(
+                'Unable to login: %s' % error_message, expected=True)
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: This duplicates code from `MedialaanIE`.

File: youtube_dl/extractor/canvas.py [hunk 10]
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
...
+
+        # If there's a ? or a # in the URL, remove them and everything after
+        clean_url = url.split('?')[0].split('#')[0].strip('/')
+        securevideo_url = clean_url + '.mssecurevideo.json'
+
+        json = self._download_json(securevideo_url, display_id)
+        # We are dealing with a '../<show>.relevant' URL
+        redirect_url = json.get('url')
+        if redirect_url:
+            return self._real_extract('https:' + redirect_url)
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: Should be delegated via `url_result`.

File: youtube_dl/extractor/canvas.py [hunk 11]
```diff
@@ -166,3 +172,142 @@ def _real_extract(self, url):
...
+        else:
+            # There is only one entry, but with an unknown key, so just get
+            # the first one
+            video_id = list(json.values())[0].get('videoid')
+            print(video_id)
+
+        return {
+            '_type': 'url_transparent',
+            'url': 'https://mediazone.vrt.be/api/v1/%s/assets/%s' % ('vrtvideo', video_id),
```
Comment (dstftw) [inline youtube_dl/extractor/canvas.py]: `vrtvideo` should be hardcoded.

File: youtube_dl/extractor/canvas.py (hunk 9 above)
Comment (mrBliss) [reply youtube_dl/extractor/canvas.py]: Thanks for the feedback! You're right, there is some code in common with `MedialaanIE`. How should I address this? Move the common code to `common.py`? (Let one inherit from the other is technically possible, but not my preference)

Comment (dstftw) [reply youtube_dl/extractor/canvas.py]: For now extract it in a base IE class.

Comment (mrBliss) [reply youtube_dl/extractor/canvas.py]: See 5ed058f for a first try.

**Comment by mrBliss** [comment]
Thanks!

**Comment by hubiqs** [comment]
@mrBliss is there any chance I can interest you in https://github.com/rg3/youtube-dl/issues/15280 ? I would very much appreciate added support for VrtNU livestreams. Thank you for all the work you've done here already!

**Pull request merged**